### ISSUE TITLE
Multi-user: share lock locations for multiple users and exclude locks from annex

### ISSUE THREAD

**Issue by christian-monch** [datalad/datalad#6593]
### What is the problem?

This problem arises if multiple users perform write operations on the same dataset concurrently (or if a single user with differing settings writes to a dataset concurrently).
When a dataset should be written to by multiple users it has to be locked. This is currently done by interprocess locks provided by `fasteners`. The provided locks are file locks, which means they require a-priori knowledge of a file-name that is used for locking (if data is written from multiple nodes, that file-named has to be resolved to a file on a shared file system). The location (or name) of the lock file has to be shared by all writers. 

There are currently two problems with locking:

1. Location sharing
2. Annex exclusion

<h2>Location sharing</h2>

The datalad configuration allows to setting a per-user location for lock files (see issue #5706). This will not suffice to protect a repository from multiple writers if those writers have configured different locations for a lock file.

To solve this problem, we should probably define a single dataset-write lock location within the dataset itself, for example, `./datalad/locks/dataset-write-lock`. Furthermore, the dataset write lock location should not be configurable.

<h2>Annex exclusion</h2>

If the lock file location is inside a dataset (see previous section for a rationale for that), it has to be excluded from the annex. That means, its path has to be added to ".gitattributes". If the configuration allows configuring lock locations to point into a dataset, those have to be excluded from the annex as well.

**Comment by mih** [comment]
Linking issues that I filed a couple of years ago on this topic.

- #4401
- #4606

I believe that there is nothing inherently "multi-user" to the core issue. I am afraid presenting it in this light will cause it to receive even less attention.

In the flip-side, this new issues is a good opportunity to close these old ones, maybe this new impulse leads to something.

**Comment by christian-monch** [comment]
I agree that the core problem is multi-writer locking. I clearly see the need for that (that's the reason it is implemented in metalad). And I am not sure to which extent we have write-locking in datalad. So, if there is a need, that has to be tackled with priority.

I just want to add that, even if we have a perfect write-locking implementation, in order to be able to synchronize processes that belong to different users, the locks have to be shared by those processes. In the case of file-based locks this sharing requires a predictable lock-file location, i.e. `$HOME/.cache/datalad/locks` would not work. I think that the configurability of lock-locations reduces the stability in a muti-user environment. We should "standardise" the location. For example: `.git/datalad/locks/dataset-write-lock` as suggested above by @yarikoptic (if that is ok with `git`).

**Comment by yarikoptic** [comment]
> For example: `.git/datalad/locks/dataset-write-lock` as suggested above by @yarikoptic (if that is ok with `git`).

AFAIK git doesn't care about `.git/datalad` so we have been using that location since forever for various purposes (e.g. populating web ui components for http://datasets.datalad.org; IIRC datalad-archive remote uses it for extraction of archives etc).

**Comment by yarikoptic** [comment]
FWIW: we do use "user-wide" lock as well ATM in downloaders to prevent multiple parallel processes asking for the (likely the same) password at the same time https://github.com/datalad/datalad/blob/HEAD/datalad/downloaders/base.py#L161 :
```
            interp_lock = InterProcessLock(
                op.join(cfg.obtain('datalad.locations.locks'),
                        'downloader-auth.lck')
            )
```
so, since comes from `cfg` could as well be not "user-wide" if changed in a specific dataset config, but it does default (https://github.com/datalad/datalad/blob/33b1af2438c42a16393e4afc9d70c7e9e9429c6d/datalad/interface/common_cfg.py#L148) to `opj(dirs.user_cache_dir, 'locks')` thus user-wide
